Scripting.md

```diff
@@ -0,0 +1,286 @@
+# Bash scripting
+### A little bit of background
+A lot of the background for bash in general is given
+[in the bash talk](CommandLine.md), so we won't retread all of that ground, but
+forgive me if I repeat some things.
+
+### Source-ing vs Executing
+There are two ways to run a shell script. The easy way through `source` and the
+slightly-more-work way of `chmod` and `./`. **There is a difference though!**
+
+`source` simply takes every line of the input file and runs it as a command on
+*your* shell. This means that all variables and side effects from the code
+remain on your shell when the script finishes executing. This can be useful but
+for most situations is not what you really want.
+
+Making a file executable and using `./` executes the commands in a *subshell*,
+meaning it's basically a different shell who's side effects do not carry back to
+your main shell. This is most exemplified in the `cd` command. It is perfectly
+reasonable to `cd` inside a script. If you call `source` on a script that uses
+`cd` you will end up wherever the script `cd`'d to. Using `./` however will
+not change your current working directory after returning.
+
+Anyways it's generally good practice to use `./`. Here's how you do it if you
+don't know:
+```bash
+chmod +x yourfile.sh
+./yourfile.sh
+```
+
+## Syntax
+#### Variables
+Setting a variable is as easy as
+```bash
+VARIABLE="foo"
+echo $VARIABLE
+```
+
+#### Variable replacements
+Replacing values in strings in bash is very simple, just takes putting the
+variable name in the string as long as you are surrounding your string with `"`.
+`'` describes a literal string however, and will not replace variables or
+expressions contained within.
+
+```bash
+echo "$VARIABLE bar"
+# foo bar
+echo '$VARIABLE bar'
+# $VARIABLE bar
+```
+
+#### Expressions in strings
+Putting an expression inside a string is similarly easy: simply surround the
+expression with `$(expression)`, or `` `expression` ``
+```bash
+echo "I'm in $(pwd)"
+echo "I'm still in `pwd`"
+```
+
+#### Brace expansion
+Bash allows you to expand things inside braces pretty reasonably:
+```bash
+echo {A, B}     # A B
+echo {A, B}.js  # A.js B.js
+echo {1..5}     # 1 2 3 4 5
+```
+
+The last type allows step size with another `..` set:
+```bash
+echo {0..50..5}
+
+#### eval
+
+Brace expansions do not work directly in strings however, but we can use the
+`eval` command to fix that:
+
+```bash
+echo "{1..5}.js"        # {1..5}.js
+eval echo "{1..5}.js"   # 1.js 2.js 3.js 4.js 5.js
+```
+
+`eval` basically runs the command twice, first through the bash processor, and
+second as the actual command. Here, bash expands the brace expansion syntax
+before sending to echo.
+
+#### Conditionals
+The `[[ ... ]]` command is an alias for `test` and returns either 0 (true) or 1
```

Comment:
`/bin/[` is an alias for `/bin/test` most of the time (sometimes it's just a copy or nearly-identical file); however, `[[ ... ]]` is always shell-specific and is not POSIX.